### PR TITLE
Add NDR

### PR DESCRIPTION
**Pull request by byeongkyu-kang**
### Summary

Added NDR to remove duplicated images in a dataset.

### How to test

```python -m unittest -v tests/test_ndr.py```
### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [ ] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: datumaro/plugins/ndr.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    Removes near-duplicated images in subset |n
+    """
+    def __init__(self, extractor, working_subset=None, duplicated_subset="duplicated", algorithm='gradient',
+            num_cut=None, over_sample='random', under_sample='uniform', seed=None, **kwargs):
+        """
+        Near-duplicated image removal
+
+        Arguments
+        ---------------
+        working_subset: str
```
Comment (zhiltsov-max) [inline datumaro/plugins/ndr.py:22]: I'm not sure if using a parameter is better than passing just a subset as an input. Also, I think, it looks more natural, when `None` run the algorithm for the whole dataset.

File: datumaro/plugins/ndr.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+            raise ValueError("working_subset == duplicated_subset")
+        if algorithm not in ('gradient'):
+            raise ValueError("Invalid algorithm name")
+        if over_sample not in ("random", "similarity"):
+            raise ValueError("Invalid over_sample")
+        if under_sample not in ("uniform", "inverse"):
+            raise ValueError("Invalid under_sample")
+        
+        self.duplicated_subset = duplicated_subset
+        self._remove(algorithm, working_subset, num_cut, over_sample, under_sample, **kwargs)
```
Comment (zhiltsov-max) [inline datumaro/plugins/ndr.py:77]: Try to make the transform lazy - it should only be applied when `__iter__` is called, look at "splitter" transforms for example. You'll also need to move the initialization of RNG into `__iter__`. Consider `__iter__` in the input dataset a heavy operation.

File: datumaro/plugins/ndr.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+                the threshold value for saving hash-collided samples.
+                larger value means more generous, i.e., saving more samples
+        Return
+        ---------------
+        None, other subsets combined with the result
+        """
+        super().__init__(extractor)
+        if seed:
+            np.random.seed(seed)
+        if len(extractor) < 1:
```
Comment (zhiltsov-max) [inline datumaro/plugins/ndr.py:58]: I don't think this situation should be considered an error.

File: datumaro/plugins/ndr.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        all_imgs = []
+        for item in working_subset:
+            if item.image.has_data:
+                having_image.append(item)
+                all_imgs.append(item.image.data)
+
+        if num_cut and num_cut > len(all_imgs):
+            raise ValueError("The number of images is smaller than the cut you want")
+        
+        if algorithm == "gradient":
```
Comment (zhiltsov-max) [inline datumaro/plugins/ndr.py:92]: Consider using an enum instead of a string constant.

File: datumaro/plugins/ndr.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+            key = all_key[ii]
+            clr = all_clr[ii]
+            if key not in clr_dict:
+                clr_dict[key] = [clr]
+                key_counter[key] = 1
+                kept_index.append(ii)
+                continue
+
+            # Hash collision: compare dot-product based feature similarity
+            max_sim = np.max(np.dot(clr_dict[key], clr) ** 50)
```
Comment (zhiltsov-max) [inline datumaro/plugins/ndr.py:135]: Please make the use of `50` more clear. It is unclear if it is just a random number, or something with a strong mathematical proof.

**Review by zhiltsov-max — CHANGES_REQUESTED** [review]
Hi, thanks for your contribution! 
Please fix formatting - avoid lines longer than 80 characters.

File: CHANGELOG.md [hunk 6]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `Icdar13/15` dataset format (<https://github.com/openvinotoolkit/datumaro/pull/96>)
 - Laziness, source caching, tracking of changes and partial updating for `Dataset` (<https://github.com/openvinotoolkit/datumaro/pull/102>)
 - `Market-1501` dataset format (<https://github.com/openvinotoolkit/datumaro/pull/108>)
+- Add NDR (<https://github.com/openvinotoolkit/datumaro/pull/113>)
```
[the remark this answers is not in the document]
Comment (byeongkyu-kang) [reply CHANGELOG.md:14]: Okay, I will update.

File: datumaro/plugins/ndr.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+class NDR(Transform):
+    """
+    Near-duplicated image removal |n
+    Removes near-duplicated images in subset |n
+    """
+    def __init__(self, extractor, working_subset=None, duplicated_subset="duplicated", algorithm='gradient',
+            num_cut=None, over_sample='random', under_sample='uniform', seed=None, **kwargs):
+        """
+        Near-duplicated image removal
```
[the remark this answers is not in the document]
Comment (byeongkyu-kang) [reply datumaro/plugins/ndr.py:18]: I think CLI will be supported by inheriting CliPlugin in another PR. Do I need to write a documentation anyway?

File: datumaro/plugins/ndr.py (hunk 1 above)
Comment (byeongkyu-kang) [reply datumaro/plugins/ndr.py:22]: I think that might return unexpected results if user has a subset like *val* or *test*.

File: datumaro/plugins/ndr.py (hunk 3 above)
Comment (byeongkyu-kang) [reply datumaro/plugins/ndr.py:58]: Okay, I will update

File: datumaro/plugins/ndr.py (hunk 2 above)
Comment (byeongkyu-kang) [reply datumaro/plugins/ndr.py:77]: Okay, I will check "splitter" and change to lazy transformation

File: datumaro/plugins/ndr.py (hunk 5 above)
Comment (byeongkyu-kang) [reply datumaro/plugins/ndr.py:135]: We do not have any experimental results for that number. The number is set arbitrarily. Then do we need to change it as a parameter?

File: datumaro/plugins/ndr.py (hunk 4 above)
Comment (byeongkyu-kang) [reply datumaro/plugins/ndr.py:92]: Could you explain more about enum?

```python
class Algo_name(Enum):
    algo_name1 = "..."
    ....

...
if algorithm == Algo_name.algo_name1
```

Do you mean creating a class by inheriting Enum like above?

File: datumaro/plugins/ndr.py (hunk 5 above)
Comment (zhiltsov-max) [reply datumaro/plugins/ndr.py:150]: I don't have any complaints against the value, but I'd better make it a named constant to clarify its use.

Comment (byeongkyu-kang) [reply datumaro/plugins/ndr.py:150]: Oh I understand. Okay, I will update

File: datumaro/plugins/ndr.py (hunk 4 above)
Comment (zhiltsov-max) [reply datumaro/plugins/ndr.py]:
```python
AlgoName = Enum('AlgoName', ['good', 'best', 'poor'])
```
Examples: https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/plugins/voc_format/format.py#L16, https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/cli/contexts/project/__init__.py#L228

File: datumaro/plugins/ndr.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply datumaro/plugins/ndr.py:20]: Let's add it in another PR then.

**Pull request merged**